### PR TITLE
Make C++ typeid accept specializations of fused types

### LINKED ISSUE #3203: typeid failed on fused types
```cython
# cython: language=c++
# cython: language_level=3str
from libcpp.typeinfo cimport type_info
from cython.operator cimport typeid
cdef fused C: 
   int 
   object 
cdef const type_info* i32 = &typeid(int) 
cdef f(C x): 
    cdef const type_info* a = &typeid(C) 
    return a  == i32 
     
def g(x): 
   return (x, f(x)) 
```

### PR DESCRIPTION
**Pull request by da-woods**
Gets the specialized type if possible from
NameNode.analyse_as_type

This does introduce a potential new bug:
```
cimport cython
from cython cimport floating

just_float = cython.fused_type(float)

cdef OK1(just_float x):
    return just_float in floating

cdef fail1(just_float x, floating y):
    return just_float in floating

cdef fail2(floating x):
    return floating in floating

def show():
    """
    >>> show()
    True
    True
    True
    True
    """
    print(OK1(1.0))
    print(fail1(1.0, 2.0))
    print(fail1[float, double](1.0, 2.0))
    print(fail2[float](1.0))
```
fail1 and fail2 work before this patch but fail with it. It isn't
clear to me if this should actually be considered a bug. It
works in both versions with `cython.floating`, which possibly
suggests analyse_as_type in AttributeNode should also be changed

Closes #3203

### CODE REVIEW

**Comment by da-woods** [comment]
Will still generate `Exception: This may never happen, please report a bug` on

```
def f():
    typeid(cython.integral)
```

(which is invalid code, but arguably should have a better error message)

**Comment by da-woods** [comment]
I've removed the try/catch if favour of a pre-check.

`x in fused_type` does seem to be tested for https://github.com/cython/cython/blob/a32a29e8aaa688e0507d374ab47e641eb1a427c4/tests/run/fused_types.pyx#L187 although I'm not sure how useful it actually is. With this change it breaks if the fused type in question is also one of the function argument. It seems like an obscure edge-case, but then this bug also seems like an obscure edge-case so maybe there isn't a huge net improvement.

**Comment by scoder** [comment]
@da-woods AFAIU, the status here is that the call to `typeid()` works now, but the type replacement at that time breaks references to the actual "global" fused type, right? What was the relation to argument types? (Fused types tend to be the argument types, that's what makes a fused function.)

Not sure how to proceed here.

**Comment by da-woods** [comment]
It's been a while since I thought about this but I think your understanding is correct.

The relation to argument types is: if you have a fused type that's used by the function (e.g. `floating` in `fail1`) then it gets specialized. If the fused type isn't used by the function (e.g. `floating` in `OK1` then it isn't specialized so it finds the global fused definition).

I don't really know how to proceed either - there's two different meanings to `cython.floating` here and they're both valid at times. Maybe I should have look at special-casing the `in` comparison? I think this might be the only time you really need the global fused definition?

**Comment by da-woods** [comment]
Right - I've updated this. Looking at it the whole handling of fused types in comparison operators is heavily special-cased anyway: https://github.com/cython/cython/blob/04d918f684404375390657d12ecb53346d66c93d/Cython/Compiler/ParseTreeTransforms.py#L3294

Therefore I've added a bit of extra special-casing to not specialize the type of operand2. I think this is almost the only time we'd want to access the "global" fused type.

**Comment by scoder** [comment]
Hmm … what about a test like `float is floating` ?

**Comment by da-woods** [comment]
In both this version (and the previous version) `float is floating` is deliberately special-cased so that `floating` gets specialized, which I think is right.

However... I think I can probably tidy up the visit_PrimaryCmpNode a bit further so leave reviewing it until that's done.

**Comment by scoder** [comment]
What about applying the specialisation explicitly in places where we _call_ `analyse_as_type()` *and* need the specialised final type? There don't seem to be all that many places like this.

We could have a new method `analyse_as_specialised_type()` which calls `analyse_as_type()` internally.

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -655,6 +655,19 @@ def analyse_as_type(self, env):
         # type, return that type, else None.
         return None
 
+    def analyse_as_specialized_type(self, env):
+        type = self.analyse_as_type(env)
+        if type.is_fused and env.fused_to_specific:
+            # while it would be nice to test "if entry.type in env.fused_to_specific"
+            # rather than try/catch this doesn't work reliably (mainly for nested fused types)
+            try:
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:663]: I know this `try:` was pointed out before in a previous iteration of this PR. It does seem necessary, unfortunately, and it isn't the only place where this pattern's used (suggesting people have found it necessary before) https://github.com/cython/cython/blob/04d918f684404375390657d12ecb53346d66c93d/Cython/Compiler/ParseTreeTransforms.py#L3344-L3349

**Comment by scoder** [comment]
Cool, that looks so much better that it feels worth the iterations.
